### PR TITLE
Extractors

### PR DESCRIPTION
**Pull request by Stormunblessed**
StreamSB fix and Jawcloud, sometimes StreamSB won't work, even with webview, I don't know exactly why.
Added GenericM3U8 for servers that are weird to extract.

### CODE REVIEW

File: app/src/main/java/com/lagradost/cloudstream3/extractors/GenericM3U8.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            )
+        )
+        val sources = mutableListOf<ExtractorLink>()
+        if (response.url.contains("m3u8")) M3u8Helper().m3u8Generation(
+            M3u8Helper.M3u8Stream(
+                response.url,
+                headers = response.headers.toMap()
+            ), true
+        )
+            .apmap { stream ->
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/GenericM3U8.kt:36]: forEatch should be used

File: app/src/main/java/com/lagradost/cloudstream3/extractors/Jawcloud.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        val doc = app.get(url).document
+        val urlString = doc.select("html body div source").attr("src")
+        val sources = mutableListOf<ExtractorLink>()
+        if (urlString.contains("m3u8"))  M3u8Helper().m3u8Generation(
+            M3u8Helper.M3u8Stream(
+                urlString,
+                headers = app.get(url).headers.toMap()
+            ), true
+        )
+            .apmap { stream ->
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/Jawcloud.kt:23]: foreatch

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt [hunk 3]
```diff
@@ -1,52 +1,140 @@
...
+                    "Origin" to mainUrl,
+                    "DNT" to "1",
+                    "Connection" to "keep-alive",
+                    "Referer" to "$mainUrl/",
+                    "Sec-Fetch-Dest" to "empty",
+                    "Sec-Fetch-Mode" to "cors",
+                    "Sec-Fetch-Site" to "cross-site",),
+            ), true
+        )
+            .apmap { stream ->
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:127]: foreach

**Comment by Stormunblessed** [comment]
forEach doesn't work there, so I use map instead, and now that I'm here I added refefer for GogoAnime extractor

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt [hunk 4]
```diff
@@ -1,52 +1,140 @@
...
+        @JsonProperty("stream_data") val streamData: StreamData,
+        @JsonProperty("status_code") val statusCode: Int,
+    )
+
+    override suspend fun getUrl(url: String, referer: String?): List<ExtractorLink>? {
+        val id = url.substringAfter("embed-").substringAfter("/e/").substringBefore("?").substringBefore(".html")
+        println(id)
+        val bytes = id.toByteArray()
+        val bytesToHex = bytesToHex(bytes)
+        val master = "$mainUrl/sources40/566d337678566f743674494a7c7c${bytesToHex}7c7c346b6767586d6934774855537c7c73747265616d7362/6565417268755339773461447c7c346133383438333436313335376136323337373433383634376337633465366534393338373136643732373736343735373237613763376334363733353737303533366236333463353333363534366137633763373337343732363536313664373336327c7c6b586c3163614468645a47617c7c73747265616d7362"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:89]: this line 89 looks very hardcoded, are you 100% sure this works for all devices and not on your session

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt [hunk 5]
```diff
@@ -1,52 +1,140 @@
...
+        @JsonProperty("backup") val backup: String,
+    )
+
+    data class Main (
+        @JsonProperty("stream_data") val streamData: StreamData,
+        @JsonProperty("status_code") val statusCode: Int,
+    )
+
+    override suspend fun getUrl(url: String, referer: String?): List<ExtractorLink>? {
+        val id = url.substringAfter("embed-").substringAfter("/e/").substringBefore("?").substringBefore(".html")
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:85]: can this be done with regex?

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt [hunk 6]
```diff
@@ -1,52 +1,140 @@
...
+        val urltext = app.get(master,
+            headers = headers,
+            allowRedirects = false
+        ).text
+        val mapped = urltext.let { parseJson<Main>(it) }
+        if (urltext.contains("m3u8")) return  M3u8Helper().m3u8Generation(
+            M3u8Helper.M3u8Stream(
+                mapped.streamData.file,
+                headers = mapOf(
+                    "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0",
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:114]: Is this specific user agent required or can you use USER_AGENT

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt [hunk 7]
```diff
@@ -1,52 +1,140 @@
...
+    override val mainUrl = "https://streamsb.net"
+}
+
+class StreamSB9 : StreamSB() {
+    override val mainUrl = "https://sbplay.one"
+}
 
-    //private val m3u8Regex = Regex(""".*?(\d*).m3u8""")
-    //private val urlRegex = Regex("""(.*?)([^/]+$)""")
+//Stolen from https://github.com/jmir1/aniyomi-extensions/blob/master/src/en/genoanime/src/eu/kanade/tachiyomi/animeextension/en/genoanime/extractors/StreamSBExtractor.kt
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:47]: They are under Apache-2.0 License so it should be good to know that impo

**Review by LagradOst — COMMENTED** [review]
just some general questions about the code

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt (hunk 4 above)
Comment (Stormunblessed) [reply app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:89]: I tested it with 2 emulators and my device, haven't got any problems, so Idk

File: app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt (hunk 6 above)
Comment (Stormunblessed) [reply app/src/main/java/com/lagradost/cloudstream3/extractors/StreamSB.kt:114]: USER_AGENT works

**Pull request merged**